// @cacheable/node-cache/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  NodeCache: () => NodeCache,
  NodeCacheErrors: () => NodeCacheErrors,
  NodeCacheStore: () => NodeCacheStore,
  default: () => index_default
});
module.exports = __toCommonJS(index_exports);
var import_utils2 = require("@cacheable/utils");
var import_hookified2 = require("hookified");

// src/store.ts
var import_utils = require("@cacheable/utils");
var import_hookified = require("hookified");
var import_keyv = __toESM(require("keyv"), 1);
var NodeCacheStore = class extends import_hookified.Hookified {
  _keyv;
  _ttl;
  _stats = new import_utils.Stats({ enabled: true });
  constructor(options) {
    super();
    this._ttl = options?.ttl;
    this._keyv = options?.store ?? new import_keyv.default();
    this._keyv.on("error", (error) => {
      this.emit("error", error);
    });
  }
  /**
   * Time to live in milliseconds.
   * @returns {number | string | undefined}
   * @readonly
   */
  get ttl() {
    return this._ttl;
  }
  /**
   * Time to live in milliseconds.
   * @param {number | string | undefined} ttl
   */
  set ttl(ttl) {
    this._ttl = ttl;
  }
  /**
   * The Keyv store instance.
   * @returns {Keyv<T>}
   * @readonly
   */
  get store() {
    return this._keyv;
  }
  /**
   * Set a key/value pair in the cache.
   * @param {string | number} key
   * @param {T} value
   * @param {number} [ttl]
   * @returns {boolean}
   */
  async set(key, value, ttl) {
    const keyValue = key.toString();
    const finalTtl = this.resolveTtl(ttl);
    await this._keyv.set(keyValue, value, finalTtl);
    this._stats.incrementKSize(keyValue);
    this._stats.incrementVSize(value);
    this._stats.incrementCount();
    return true;
  }
  /**
   * Set multiple key/value pairs in the cache.
   * @param {PartialNodeCacheItem[]} list
   * @returns {void}
   */
  async mset(list) {
    for (const item of list) {
      const keyValue = item.key.toString();
      const finalTtl = this.resolveTtl(item.ttl);
      await this._keyv.set(keyValue, item.value, finalTtl);
      this._stats.incrementKSize(keyValue);
      this._stats.incrementVSize(item.value);
      this._stats.incrementCount();
    }
  }
  /**
   * Get a value from the cache.
   * @param {string | number} key
   * @returns {any | undefined}
   */
  async get(key) {
    const result = await this._keyv.get(key.toString());
    if (result === void 0) {
      this._stats.incrementMisses();
    } else {
      this._stats.incrementHits();
    }
    return result;
  }
  /**
   * Get multiple values from the cache.
   * @param {Array<string | number>} keys
   * @returns {Record<string, any | undefined>}
   */
  async mget(keys) {
    const result = {};
    for (const key of keys) {
      const value = await this._keyv.get(key.toString());
      if (value === void 0) {
        this._stats.incrementMisses();
      } else {
        this._stats.incrementHits();
      }
      result[key.toString()] = value;
    }
    return result;
  }
  /**
   * Delete a key from the cache.
   * @param {string | number} key
   * @returns {boolean}
   */
  async del(key) {
    const keyValue = key.toString();
    const value = await this._keyv.get(keyValue);
    const result = await this._keyv.delete(keyValue);
    if (result) {
      this._stats.decreaseKSize(keyValue);
      if (value !== void 0) {
        this._stats.decreaseVSize(value);
      }
      this._stats.decreaseCount();
    }
    return result;
  }
  /**
   * Delete multiple keys from the cache.
   * @param {Array<string | number>} keys
   * @returns {boolean}
   */
  async mdel(keys) {
    for (const key of keys) {
      const keyValue = key.toString();
      const value = await this._keyv.get(keyValue);
      this._stats.decreaseKSize(keyValue);
      if (value !== void 0) {
        this._stats.decreaseVSize(value);
      }
      this._stats.decreaseCount();
    }
    return this._keyv.delete(keys.map((key) => key.toString()));
  }
  /**
   * Clear the cache.
   * @returns {void}
   */
  async clear() {
    await this._keyv.clear();
    this._stats.resetStoreValues();
  }
  /**
   * Set the TTL of an existing key in the cache.
   * @param {string | number} key
   * @param {number | string} [ttl]
   * @returns {boolean}
   */
  async setTtl(key, ttl) {
    const item = await this._keyv.get(key.toString());
    if (item) {
      const finalTtl = this.resolveTtl(ttl);
      await this._keyv.set(key.toString(), item, finalTtl);
      return true;
    }
    return false;
  }
  /**
   * Get a key from the cache and delete it. If it does exist it will get the value and delete the item from the cache.
   * @param {string | number} key
   * @returns {T | undefined}
   */
  async take(key) {
    const keyValue = key.toString();
    const result = await this._keyv.get(keyValue);
    if (result !== void 0) {
      await this._keyv.delete(keyValue);
      this._stats.incrementHits();
      this._stats.decreaseKSize(keyValue);
      this._stats.decreaseVSize(result);
      this._stats.decreaseCount();
    } else {
      this._stats.incrementMisses();
    }
    return result;
  }
  /**
   * Gets the stats of the cache
   * @returns {NodeCacheStats} the stats of the cache
   */
  getStats() {
    return {
      keys: this._stats.count,
      hits: this._stats.hits,
      misses: this._stats.misses,
      ksize: this._stats.ksize,
      vsize: this._stats.vsize
    };
  }
  /**
   * Flush the stats.
   * @returns {void}
   */
  flushStats() {
    this._stats = new import_utils.Stats({ enabled: true });
    this.emit("flush_stats");
  }
  /**
   * Disconnect from the cache.
   * @returns {void}
   */
  async disconnect() {
    await this._keyv.disconnect();
  }
  /**
   * Resolve the TTL to milliseconds.
   * @param {number | string | undefined} ttl
   * @returns {number | undefined}
   */
  resolveTtl(ttl) {
    const effectiveTtl = ttl ?? this._ttl;
    if (effectiveTtl === void 0) {
      return void 0;
    }
    if (typeof effectiveTtl === "string") {
      return (0, import_utils.shorthandToMilliseconds)(effectiveTtl);
    }
    if (effectiveTtl === 0) {
      return void 0;
    }
    return effectiveTtl;
  }
};

// src/index.ts
var NodeCacheErrors = /* @__PURE__ */ ((NodeCacheErrors2) => {
  NodeCacheErrors2["ECACHEFULL"] = "Cache max keys amount exceeded";
  NodeCacheErrors2["EKEYTYPE"] = "The key argument has to be of type `string` or `number`. Found: `__key`";
  NodeCacheErrors2["EKEYSTYPE"] = "The keys argument has to be an array.";
  NodeCacheErrors2["ETTLTYPE"] = "The ttl argument has to be a number or a string for shorthand ttl.";
  return NodeCacheErrors2;
})(NodeCacheErrors || {});
var NodeCache = class extends import_hookified2.Hookified {
  options = {
    stdTTL: 0,
    checkperiod: 600,
    useClones: true,
    deleteOnExpire: true,
    maxKeys: -1
  };
  store = /* @__PURE__ */ new Map();
  _stats = new import_utils2.Stats({ enabled: true });
  intervalId = 0;
  constructor(options) {
    super({ throwOnHookError: false });
    if (options) {
      this.options = { ...this.options, ...options };
    }
    this.startInterval();
  }
  /**
   * Sets a key value pair. It is possible to define a ttl (in seconds). Returns true on success.
   *
   * TTL behavior:
   * - `ttl > 0`: cache expires after the given number of seconds
   * - `ttl === 0`: cache indefinitely (overrides stdTTL)
   * - `ttl < 0`: store the value but it expires immediately on next access (matches original node-cache behavior)
   * - `ttl` omitted/undefined: use `stdTTL` from options (0 = unlimited if stdTTL is 0 or not set)
   * - `ttl` as string: shorthand format like '1h', '30s', '5m', '2d'
   *
   * @param {string | number} key - it will convert the key to a string
   * @param {T} value
   * @param {number | string} [ttl] - TTL in seconds. 0 = unlimited, negative = expires immediately, string = shorthand format
   * @returns {boolean} true on success
   * @throws {Error} If the `key` or `ttl` is of an invalid type.
   */
  set(key, value, ttl) {
    if (typeof key !== "string" && typeof key !== "number") {
      throw this.createError("The key argument has to be of type `string` or `number`. Found: `__key`" /* EKEYTYPE */, key);
    }
    if (ttl !== void 0 && typeof ttl !== "number" && typeof ttl !== "string") {
      throw this.createError("The ttl argument has to be a number or a string for shorthand ttl." /* ETTLTYPE */, this.formatKey(key));
    }
    const keyValue = this.formatKey(key);
    let expirationTimestamp = 0;
    if (this.isNegativeTtl(ttl)) {
      expirationTimestamp = Math.max(
        1,
        this.getExpirationTimestamp(
          typeof ttl === "string" ? Number(ttl) : ttl
        )
      );
    } else if (ttl !== void 0 && (typeof ttl === "string" || ttl > 0)) {
      expirationTimestamp = this.resolveExpiration(ttl);
    } else if (ttl === void 0 && this.options.stdTTL !== void 0 && this.options.stdTTL !== 0) {
      expirationTimestamp = this.resolveExpiration(this.options.stdTTL);
    }
    if (this.options.maxKeys) {
      const { maxKeys } = this.options;
      if (maxKeys > -1 && this.store.size >= maxKeys) {
        throw this.createError("Cache max keys amount exceeded" /* ECACHEFULL */, this.formatKey(key));
      }
    }
    this.store.set(keyValue, {
      key: keyValue,
      value,
      ttl: expirationTimestamp
    });
    this.emit("set", keyValue, value, expirationTimestamp);
    this._stats.incrementKSize(keyValue);
    this._stats.incrementVSize(value);
    this._stats.setCount(this.store.size);
    return true;
  }
  /**
   * Sets multiple key val pairs. It is possible to define a ttl (seconds). Returns true on success.
   *
   * Each item follows the same TTL behavior as `set()`:
   * - Positive TTL: expires after the given seconds
   * - `0`: cache indefinitely
   * - Negative TTL: stored but expires immediately on next access
   * - Omitted: uses `stdTTL` from options
   *
   * @param {PartialNodeCacheItem<T>[]} data an array of key value pairs with optional ttl
   * @returns {boolean} true on success
   * @throws {Error} If `data` is not an array, or if any item has an invalid key or ttl type.
   */
  mset(data) {
    if (!Array.isArray(data)) {
      throw this.createError("The keys argument has to be an array." /* EKEYSTYPE */);
    }
    for (const item of data) {
      this.set(item.key, item.value, item.ttl);
    }
    return true;
  }
  /**
   * Gets a saved value from the cache. Returns a undefined if not found or expired. If the value was found it returns the value.
   * @param {string | number} key if the key is a number it will convert it to a string
   * @returns {T} the value or undefined
   */
  get(key) {
    const result = this.store.get(this.formatKey(key));
    if (result) {
      if (result.ttl > 0) {
        if (result.ttl < Date.now()) {
          if (this.options.deleteOnExpire) {
            this.del(key);
          }
          this._stats.incrementMisses();
          this.emit("expired", this.formatKey(key), result.value);
          return void 0;
        }
        this._stats.incrementHits();
        if (this.options.useClones) {
          return this.clone(result.value);
        }
        return result.value;
      }
      this._stats.incrementHits();
      if (this.options.useClones) {
        return this.clone(result.value);
      }
      return result.value;
    }
    this._stats.incrementMisses();
    return void 0;
  }
  /**
   * Gets multiple saved values from the cache. Returns an empty object {} if not found or expired.
   * If the value was found it returns an object with the key value pair.
   * @param {Array<string | number} keys an array of keys
   * @returns {Record<string, T | undefined>} an object with the key as a property and the value as the value
   */
  mget(keys) {
    const result = {};
    for (const key of keys) {
      const value = this.get(key);
      if (value) {
        result[this.formatKey(key)] = value;
      }
    }
    return result;
  }
  /**
   * Get the cached value and remove the key from the cache. Equivalent to calling get(key) + del(key).
   * Useful for implementing single use mechanism such as OTP, where once a value is read it will become obsolete.
   * @param {string | number} key
   * @returns {T | undefined} the value or undefined
   */
  take(key) {
    const result = this.get(key);
    if (result) {
      this.del(key);
      if (this.options.useClones) {
        return this.clone(result);
      }
      return result;
    }
    return void 0;
  }
  /**
   * Delete a key. Returns the number of deleted entries. A delete will never fail.
   * @param {string | number | Array<string | number>} key if the key is a number it will convert it to a string. if an array is passed it will delete all keys in the array.
   * @returns {number} if it was successful it will return the count that was deleted
   */
  del(key) {
    if (Array.isArray(key)) {
      return this.mdel(key);
    }
    const result = this.store.get(this.formatKey(key));
    if (result) {
      const keyValue = this.formatKey(key);
      this.store.delete(keyValue);
      this.emit("del", keyValue, result.value);
      this._stats.decreaseKSize(keyValue);
      this._stats.decreaseVSize(result.value);
      this._stats.setCount(this.store.size);
      return 1;
    }
    return 0;
  }
  /**
   * Delete all keys in Array that exist. Returns the number of deleted entries.
   * @param {Array<string | number>} keys an array of keys
   * @returns {number} the count of deleted keys
   */
  mdel(keys) {
    let result = 0;
    for (const key of keys) {
      result += this.del(key);
    }
    return result;
  }
  /**
   * Redefine the ttl of a key. Returns true if the key has been found and changed.
   * Otherwise returns false. If the ttl-argument isn't passed the default-TTL will be used.
   *
   * TTL behavior:
   * - `ttl > 0`: key expires after the given number of seconds
   * - `ttl === 0`: key lives indefinitely (overrides stdTTL)
   * - `ttl < 0`: key expires immediately on next access (matches original node-cache behavior)
   * - `ttl` omitted/undefined: use `stdTTL` from options (0 = unlimited if stdTTL is 0 or not set)
   * - `ttl` as string: shorthand format like '1h', '30s', '5m', '2d'
   *
   * @param {string | number} key if the key is a number it will convert it to a string
   * @param {number | string} [ttl] TTL in seconds. 0 = unlimited, negative = expires immediately, string = shorthand format
   * @returns {boolean} true if the key has been found and changed. Otherwise returns false.
   * @throws {Error} If the `ttl` is of an invalid type (must be a number or string).
   */
  ttl(key, ttl) {
    if (ttl !== void 0 && typeof ttl !== "number" && typeof ttl !== "string") {
      throw this.createError("The ttl argument has to be a number or a string for shorthand ttl." /* ETTLTYPE */, this.formatKey(key));
    }
    const result = this.store.get(this.formatKey(key));
    if (result) {
      if (this.isNegativeTtl(ttl)) {
        result.ttl = Math.max(
          1,
          this.getExpirationTimestamp(
            typeof ttl === "string" ? Number(ttl) : ttl
          )
        );
      } else if (ttl !== void 0 && (typeof ttl === "string" || ttl > 0)) {
        result.ttl = this.resolveExpiration(ttl);
      } else if (ttl === 0) {
        result.ttl = 0;
      } else if (this.options.stdTTL !== void 0 && this.options.stdTTL !== 0) {
        result.ttl = this.resolveExpiration(this.options.stdTTL);
      } else {
        result.ttl = 0;
      }
      this.store.set(this.formatKey(key), result);
      return true;
    }
    return false;
  }
  /**
   * Receive the ttl of a key.
   * @param {string | number} key if the key is a number it will convert it to a string
   * @returns {number | undefined} 0 if this key has no ttl, undefined if this key is not in the cache,
   * a timestamp in ms representing the time at which this key will expire
   */
  getTtl(key) {
    const result = this.store.get(this.formatKey(key));
    if (result) {
      if (result.ttl === 0) {
        return 0;
      }
      return result.ttl;
    }
    return void 0;
  }
  /**
   * Returns an array of all existing keys. [ "all", "my", "keys", "foo", "bar" ]
   * @returns {string[]} an array of all keys
   */
  keys() {
    const result = [];
    for (const key of this.store.keys()) {
      result.push(key);
    }
    return result;
  }
  /**
   * Returns boolean indicating if the key is cached.
   * @param {string | number} key if the key is a number it will convert it to a string
   * @returns {boolean} true if the key is cached
   */
  has(key) {
    return this.store.has(this.formatKey(key));
  }
  /**
   * Gets the stats of the cache
   * @returns {NodeCacheStats} the stats of the cache
   */
  getStats() {
    const stats = {
      keys: this._stats.count,
      hits: this._stats.hits,
      misses: this._stats.misses,
      ksize: this._stats.ksize,
      vsize: this._stats.vsize
    };
    return stats;
  }
  /**
   * Flush the whole data.
   * @returns {void}
   */
  flushAll() {
    this.store.clear();
    this.flushStats();
    this.emit("flush");
  }
  /**
   * Flush the stats.
   * @returns {void}
   */
  flushStats() {
    this._stats = new import_utils2.Stats({ enabled: true });
    this.emit("flush_stats");
  }
  /**
   * Close the cache. This will clear the interval timeout which is set on check period option.
   * @returns {void}
   */
  close() {
    this.stopInterval();
  }
  /**
   * Get the interval id
   * @returns {number | NodeJS.Timeout} the interval id
   */
  getIntervalId() {
    return this.intervalId;
  }
  startInterval() {
    if (this.options.checkperiod && this.options.checkperiod > 0) {
      const checkPeriodinSeconds = this.options.checkperiod * 1e3;
      this.intervalId = setInterval(() => {
        this.checkData();
      }, checkPeriodinSeconds).unref();
      return;
    }
    this.intervalId = 0;
  }
  stopInterval() {
    if (this.intervalId !== 0) {
      clearInterval(this.intervalId);
      this.intervalId = 0;
    }
  }
  formatKey(key) {
    return key.toString();
  }
  // biome-ignore lint/suspicious/noExplicitAny: type format
  clone(value) {
    if (value === null || value === void 0) {
      return value;
    }
    if (typeof value !== "object") {
      return value;
    }
    return structuredClone(value);
  }
  getExpirationTimestamp(ttlInSeconds) {
    if (typeof ttlInSeconds === "string") {
      return (0, import_utils2.shorthandToTime)(ttlInSeconds);
    }
    const currentTimestamp = Date.now();
    const ttlInMilliseconds = ttlInSeconds * 1e3;
    const expirationTimestamp = currentTimestamp + ttlInMilliseconds;
    return expirationTimestamp;
  }
  /**
   * Resolves a TTL value to an expiration timestamp, returning 0 (unlimited) if the
   * resolved timestamp is not in the future (e.g. "0s" or a zero-duration string).
   */
  resolveExpiration(ttl) {
    const timestamp = this.getExpirationTimestamp(ttl);
    if (timestamp <= Date.now()) {
      return 0;
    }
    return timestamp;
  }
  /**
   * Checks whether a TTL value is negative. Handles both numbers and
   * purely numeric strings (e.g. "-1").
   */
  isNegativeTtl(ttl) {
    if (typeof ttl === "number") {
      return ttl < 0;
    }
    if (typeof ttl === "string") {
      const num = Number(ttl);
      if (!Number.isNaN(num) && num < 0) {
        return true;
      }
    }
    return false;
  }
  checkData() {
    for (const [key, value] of this.store.entries()) {
      if (value.ttl > 0 && value.ttl < Date.now()) {
        if (this.options.deleteOnExpire) {
          this.del(key);
        }
        this.emit("expired", this.formatKey(key), value.value);
      }
    }
  }
  createError(errorCode, key) {
    let error = errorCode;
    if (key) {
      error = error.replace("__key", key);
    }
    return new Error(error);
  }
};
var index_default = NodeCache;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NodeCache,
  NodeCacheErrors,
  NodeCacheStore
});
/* v8 ignore next -- @preserve */
